File: app/src/main/java/com/practice/todo/storage/sharedpreferences/SpDel.java
package com.practice.todo.storage.sharedpreferences;

import android.content.Context;
import android.content.SharedPreferences;

import com.practice.todo.App;

/**
 * 使用安卓的SharedPreference来存储用户是否第一次登陆。
 */
public class SpDel {

    private static final String APP_SP = "app_sp";
    private static final String IS_FIRST_RUN = "is_first_run";

    public static void setIsFirstRun(boolean isFirstRun) {
        SharedPreferences sp = App.instance.getSharedPreferences(APP_SP, Context.MODE_PRIVATE);
        sp.edit().putBoolean(IS_FIRST_RUN, isFirstRun).apply();
    }

    public static boolean isFirstRun() {
        SharedPreferences sp = App.instance.getSharedPreferences(APP_SP, Context.MODE_PRIVATE);
        return sp.getBoolean(IS_FIRST_RUN, true);
    }

}

File: app/src/main/java/com/practice/todo/storage/sharedpreferences/SpDel.kt
package com.practice.todo.storage.sharedpreferences

import android.content.Context
import android.content.SharedPreferences
import com.practice.todo.App
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

/**
 * SharedPreferences property delegate.
 */
open class SpDel<T>(
    private val spFileName: String,
    private val key: String,
    private val defaultValue: T
) : ReadWriteProperty<Any?, T> {

    private val sp: SharedPreferences
        get() = App.instance.getSharedPreferences(spFileName, Context.MODE_PRIVATE)

    override fun getValue(thisRef: Any?, property: KProperty<*>) = findPreference(key, defaultValue)

    override fun setValue(thisRef: Any?, property: KProperty<*>, value: T) =
        putPreference(key, value)

    private fun <T> findPreference(name: String, default: T): T = with(sp) {
        val res: Any = when (default) {
            is Long -> getLong(name, default)
            is String -> getString(name, default)
            is Int -> getInt(name, default)
            is Boolean -> getBoolean(name, default)
            is Float -> getFloat(name, default)
            else -> throw IllegalArgumentException("This type can not be saved into Preferences")
        }
        @Suppress("UNCHECKED_CAST")
        res as T
    }

    private fun <U> putPreference(name: String, value: U) = with(sp.edit()) {
        when (value) {
            is Long -> putLong(name, value)
            is String -> putString(name, value)
            is Int -> putInt(name, value)
            is Boolean -> putBoolean(name, value)
            is Float -> putFloat(name, value)
            else -> throw IllegalArgumentException("This type can not be saved into Preferences")
        }.apply()
    }
}

